narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright 2020 Red Hat, Inc, and individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package me.snowdrop.boot.narayana.core.jdbc.lrco;
+
+import java.sql.Array;
+import java.sql.Blob;
+import java.sql.CallableStatement;
+import java.sql.Clob;
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.NClob;
+import java.sql.PreparedStatement;
+import java.sql.SQLClientInfoException;
+import java.sql.SQLException;
+import java.sql.SQLWarning;
+import java.sql.SQLXML;
+import java.sql.Savepoint;
+import java.sql.Statement;
+import java.sql.Struct;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Executor;
+
+import javax.sql.ConnectionEvent;
+import javax.sql.ConnectionEventListener;
+import javax.sql.StatementEventListener;
+import javax.sql.XAConnection;
+import javax.transaction.xa.XAResource;
+
+public class LrcoXAConnection implements XAConnection {
+
+    private final Connection physicalConnection;
+    private volatile Connection handleConnection;
+    private final List<ConnectionEventListener> eventListeners;
+
+    public LrcoXAConnection(Connection connection) throws SQLException {
+        this.physicalConnection = connection;
+        this.eventListeners = new ArrayList<>();
+    }
+
+    @Override
+    public XAResource getXAResource() throws SQLException {
+        return new LrcoXAResource(this.physicalConnection);
+    }
+
+    @Override
+    public Connection getConnection() throws SQLException {
+        Connection lastHandleConnection = this.handleConnection;
```

Comment:
Could you explain a bit this method? What's the purpose of the PooledJdbcConnection class? Other than close it seems to always delegate to the physical connection.

Reply:
That's the contract from JDBC API:

Creates and returns a <code>Connection</code> object that is a handle
for the physical connection that
this <code>PooledConnection</code> object represents.
The connection pool manager calls this method when an application has
called the method <code>DataSource.getConnection</code> and there are
no <code>PooledConnection</code> objects available. See the
{@link PooledConnection interface description} for more information.